aiosmtpd/tests/test_starttls.py

```diff
@@ -1,236 +1,301 @@
+
+
+@pytest.mark.usefixtures("tls_controller")
+class TestStartTLS:
+    def test_help_starttls(self, tls_controller, client):
+        resp = client.docmd("HELP STARTTLS")
+        assert resp == S.S250_SYNTAX_STARTTLS
+
+    def test_starttls_arg(self, tls_controller, client):
+        resp = client.docmd("STARTTLS arg")
+        assert resp == S.S501_SYNTAX_STARTTLS
+
+    @pytest.mark.handler_data(class_=ReceivingHandler)
+    def test_starttls(self, tls_controller, client):
+        sender = "sender@example.com"
+        recipients = ["rcpt1@example.com"]
+        code, _ = client.ehlo("example.com")
+        assert code == 250
+        assert "starttls" in client.esmtp_features
+        resp = client.starttls()
+        assert resp == S.S220_READY_TLS
+        client.send_message(MIMEText("hi"), sender, recipients)
+        handler: ReceivingHandler = tls_controller.handler
+        assert len(handler.box) == 1
+        assert handler.box[0].mail_from == sender
+        assert handler.box[0].rcpt_tos == recipients
+
+    def test_starttls_quit(self, tls_controller, client):
+        code, _ = client.ehlo("example.com")
+        assert code == 250
+        resp = client.starttls()
+        assert resp == S.S220_READY_TLS
+        resp = client.quit()
+        assert resp == S.S221_BYE
+        client.close()
+
+    @pytest.mark.handler_data(class_=HandshakeFailingHandler)
+    def test_failed_handshake(self, client):
+        code, _ = client.ehlo("example.com")
+        assert code == 250
+        resp = client.starttls()
+        assert resp == S.S220_READY_TLS
+        resp = client.mail("sender@example.com")
+        assert resp == S.S554_LACK_SECURITY
+        resp = client.rcpt("rcpt@example.com")
+        assert resp == S.S554_LACK_SECURITY
+
+    def test_tls_bad_syntax(self, client):
+        code, _ = client.ehlo("example.com")
+        assert code == 250
+        resp = client.docmd("STARTTLS", "TRUE")
+        assert resp == S.S501_SYNTAX_STARTTLS
+
+    def test_help_after_starttls(self, client):
+        resp = client.docmd("HELP")
+        assert resp == S.S250_SUPPCMD_TLS
+
+    def test_helo_starttls(self, tls_controller, client):
+        resp = client.helo("example.com")
+        assert resp == S.S250_FQDN
+        # Entering portion of code where hang is possible (upon assertion fail), so
+        # we must wrap with "try..finally".
+        try:
+            resp = client.docmd("STARTTLS")
+            assert resp == S.S220_READY_TLS
+        finally:
+            tls_controller.stop()
+
+
+class EOFingHandler:
+    ssl_existed = None
+    result = None
+
+    async def handle_NOOP(self, server: Server, session: Sess_, envelope, arg):
+        self.ssl_existed = session.ssl is not None
+        self.result = server.eof_received()
+        return "250 OK"
+
+
+class TestTLSEnding:
+
+    @pytest.mark.handler_data(class_=EOFingHandler)
+    def test_eof_received(self, tls_controller, client):
```

Comment:
This is where I finally (apparently) conquered the dreaded "intermittent coverage failure" issue. And successfully removed `# pragma: nomswin` for that method as well!

Obligatory Knocking on Wood follows...